### DISCUSSION TITLE
Should my team use Copilot for Business?

### DISCUSSION THREAD

**Discussion by blackgirlbytes** [community/community#49422] [Discover: GitHub Best Practices]
In February 2023, GitHub announced that Copilot for Business is widely available to Free, Team, and GitHub Enterprise Cloud customers. While Copilot for Individuals provides individual users access to Copilot, Copilot for Business makes it easier for companies to purchase and manage licenses for their entire team. If you're considering investing in GitHub Copilot for your team, here's what you need to know.

## What is GitHub Copilot? 🤖
It is a cloud-based artificial intelligence tool developed by GitHub and OpenAI to assist users of Visual Studio, Visual Studio Code, Neovim, and JetBrains by suggesting lines of code and whole functions instantly. You can learn more about GitHub Copilot from the following resources:

* [An introduction to GitHub Copilot](https://github.com/orgs/community/discussions/39086)
* [GitHub Copilot Research Recitation](https://github.blog/2021-06-30-github-copilot-research-recitation/)
* [How GitHub Copilot helps improve developer productivity](https://github.blog/2022-07-14-research-how-github-copilot-helps-improve-developer-productivity/)
* [8 things you didn’t know you could do with GitHub Copilot](https://github.blog/2022-09-14-8-things-you-didnt-know-you-could-do-with-github-copilot/)

## The friction between software development teams and businesses 💼

Businesses need to make money to survive, pay their employees, and continue paying for the technology that supports their applications. Investors expect a return on investment, and customers expect features that meet their needs. In an ideal world, software developers would have the time and resources to write, test, refactor, and deploy code at their own pace. But the reality is that businesses need features delivered quickly. This can lead to pressure on software developers to move faster, but to the non-developers it can feel like software developers aren’t moving fast enough.

### Why are software developers taking so long to finish features and fix bugs? ⏰

Software development is hard. It’s more than writing code, but many times people, including software engineers, don’t realize that. Sometimes as software developers, we struggle with:
* Estimating the time it will take to complete a task
* Envisioning all the problems they may encounter while coding
* Requesting extra time to understand the complexity of an issue before agreeing to an unreasonable deadline

Many times in my career, business-centric coworkers like salespeople or product managers have told me, “Here’s the feature we want, but it shouldn’t take you too long.” Spoiler alert: I hate to admit it, but it always takes me longer than expected. Sometimes, it may seem like I just have to update one word or change one picture. However, the request might mean I have to venture into an unfamiliar part of the codebase, learn a new tool, or make a change that could potentially disrupt the entire system. I come from the startup world where we’ve adopted the “Move fast and break things” motto. Realistically, my software development process often looks like this:

* Researching possibilities
* Designing the feature
* Reading documentation
* Starting to code while referring to documentation
* Realizing that the initial plan won't work and starting over
* Adding workarounds to meet the deadline
* Rushed testing and deployment
* Fixing bugs after the code breaks in production

This cycle of coding, testing, and fixing can lead to burnout and decreased motivation to continue building skills and contributing to the company's success.

### How to keep software developers happy at your company 😇
Software developers want to feel valued and have the opportunity to continue building their skills.

Helping software developers improve their skills is a compelling way for companies to demonstrate their appreciation for developers. To improve their coding skills, software engineers consume educational content, build side projects, attend conferences, and more. However, with a heavy workload, finding time to do anything outside of work-related tasks can be challenging.  Many companies provide a percentage of learning time as a benefit for software engineers, but it’s hard to use this benefit when the engineer’s learning time is not accounted for within project deadlines.

In addition to factoring in a better schedule to balance workload and learning, companies can help reclaim developers upskilling time with Copilot for Business.

### How GitHub Copilot can help your team 🤝

GitHub Copilot helps software developers write code faster by reducing the time spent sifting through documentation, writing boilerplate code, and trying to remember syntax. With GitHub Copilot, software developers can have more time to tap into their passion for coding. Overall, this helps businesses retain investors, customers, and most importantly, software developers. (Yes, you can always replace software developers, but software developers with legacy knowledge of the codebase are invaluable to a company’s success.)

#### Less time reading documentation 📚

In the example below, [Chris Sev](https://twitter.com/chris__sev) explains that instead of spending too much time reading documentation and context-switching, GitHub Copilot led him in the right direction.
<div>
<blockquote class="twitter-tweet"><p lang="en" dir="ltr">copilot just saved me 547 minutes of reading stripe docs <a href="https://t.co/f1JuRYq32i">pic.twitter.com/f1JuRYq32i</a></p>&mdash; Chris Sev (@chris__sev) <a href="https://twitter.com/chris__sev/status/1623452869516042241?ref_src=twsrc%5Etfw">February 8, 2023</a></blockquote> 
</div>

#### Less time writing boilerplate code 🧐

In the example below, [Kelsey Hightower](https://twitter.com/kelseyhightower) used GitHub Copilot to handle writing boilerplate code. 

<div>
<blockquote class="twitter-tweet"><p lang="en" dir="ltr">After spending a few minutes with GitHub&#39;s Copilot I can see the potential. It&#39;s autocomplete with way more context about what you&#39;re working on.<br><br>Just a simple web app but Copilot was able to generate 95% of the code for me. Makes you realize just how much boilerplate we write. <a href="https://t.co/rau501Bba6">pic.twitter.com/rau501Bba6</a></p>&mdash; Kelsey Hightower (@kelseyhightower) <a href="https://twitter.com/kelseyhightower/status/1617684526448402434?ref_src=twsrc%5Etfw">January 24, 2023</a></blockquote>
</div>

#### Less time recalling syntax 🧠

We all forget syntax sometimes. Memorizing code is not the best use of anyone’s time. In the example below, GitHub Copilot helps [Lisa Reiber](https://twitter.com/lisa_reiber) focus on the logic rather than the syntax. 

<div>
<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Take on learning with <a href="https://twitter.com/hashtag/githubcopilot?src=hash&amp;ref_src=twsrc%5Etfw">#githubcopilot</a>: <br><br>for standard syntax it helps to focus on the topic I’m learning (sending data from data lake to data warehouse) instead of searching for the 50th time “if name main syntax” 🎉 <br><br>it’s even a bit faster than typing<a href="https://twitter.com/hashtag/dezoomcamp?src=hash&amp;ref_src=twsrc%5Etfw">#dezoomcamp</a> <a href="https://twitter.com/hashtag/vscode?src=hash&amp;ref_src=twsrc%5Etfw">#vscode</a> <a href="https://t.co/Oz607dVPDr">pic.twitter.com/Oz607dVPDr</a></p>&mdash; Lisa Reiber (@lisa_reiber) <a href="https://twitter.com/lisa_reiber/status/1621529827416375296?ref_src=twsrc%5Etfw">February 3, 2023</a></blockquote> 
</div>

### How GitHub Copilot is useful for me – your underrepresented junior engineer ⭐️

Beyond productivity, GitHub Copilot helps provide me with psychological safety. Psychological safety is the belief that you can speak up, ask questions, make mistakes, and contribute without fear of punishment or humiliation. When employees feel safe to learn, ask questions, contribute, and lead, it boosts the team’s and individual’s overall confidence, morale, and performance. Unfortunately, psychological safety is not always common at work, especially in the tech industry, especially for minorities.

I’m a young, black immigrant woman with a non-traditional career path as a developer. I don’t see many people that look like me in my field. I’m passionate about coding, but my first two years working as a Software Engineer were stressful. After I graduated from a coding bootcamp in 2018, I excitedly landed my first software engineering role, but the excitement gradually transformed into anxiety. Asking my teammates for help often led to my teammates questioning my abilities, saying things like:
 * “I thought you would already know this.” 
* “One day, you will learn how to think like a software engineer.” 
* “I know you’re trying to prove yourself as a minority.” 
I felt so much self-doubt and anxiety from these experiences that it took me hours to open Visual Studio Code. I almost quit coding. 

Discovering GitHub Copilot restored my love for coding. I still ask my teammates questions because I am a collaborative engineer. However, GitHub Copilot acts as my non-judgemental pair programmer, guiding me in the right direction and providing multiple suggestions as I use an API or tool for the first time.

To me, Copilot for Business can be an investment in empowering your team’s newest developer. 

### How GitHub Copilot makes coding more accessible 💪

GitHub Copilot improves the developer experience for engineers with challenges we often overlook such as visual impairment or physical dexterity. The GitHub Next team is a group of researchers and engineers focused on investigating the future of software development, and they developed a hands-free voice-activated AI programmer called “Hey, GitHub.”

### Are there any privay concerns I should be aware of? 🕵️‍♂️
We’ve addressed the privacy and data concerns by implementing the following features: 
* With Copilot for Business, GitHub won’t retain code snippets, store, or share your code regardless if the data is from public repositories, private repositories, non-GitHub repositories, or local files.
* Additionally, we built a [filter](https://docs.github.com/en/copilot/configuring-github-copilot/configuring-github-copilot-settings-on-githubcom) to help detect and suppress GitHub Copilot suggestions that contain code that matches public code on GitHub. It checks code suggestions with its surrounding code for matches or near matches (ignoring whitespace) against public code on GitHub of about 150 characters. If there is a match, the suggestion will not be shown to you.
* We are currently building a feature that will provide a reference for suggestions that resemble public code on GitHub so that you can make a more informed decision about whether and how to use that code, as well as explore and learn how that code is used in other projects.
* We built an AI-based vulnerability prevention system that blocks insecure coding patterns in real-time to make GitHub Copilot suggestions more secure. Our model targets the most common vulnerable coding patterns, including hardcoded credentials, SQL injections, and path injections.

To learn more about the improvements we’ve made, take a look at the following resources:
* [GitHub Copilot for Business is now available](https://github.blog/2023-02-14-github-copilot-for-business-is-now-available/)
* [GitHub Copilot Frequently Asked Questions](https://github.com/features/copilot)

### Sign up 🎉
It’s up to you to decide if Copilot for business is right for your team, but if you’re excited to empower your team, [sign up here](https://github.com/features/copilot)!

### Any questions? 💡 
If you have questions, feel free to ask in our [Community discussion board](https://github.com/orgs/community/discussions)

**Comment by ngocduysdtq** [comment]
I am a customer of GitHub Enterprise and I am currently looking into the GitHub Copilot feature to use in our company's development projects. However, I haven't found specific information on whether this feature is included in the GitHub Enterprise version or not.

Therefore, I would like to ask if the GitHub Copilot feature is included in the GitHub Enterprise version or not. If so, how can we activate this feature?

**Reply by blackgirlbytes** [reply]
At the moment, I don't believe it's included in GitHub Enterprise. You would have to purchase it through github.com/features/copilot and assign a license to people within your organization.

**Reply by ngocduysdtq** [reply]
thankyou

**Comment by JLeander23** [comment]
> In February 2023, GitHub announced that Copilot for Business is widely available to Free, Team, and GitHub Enterprise Cloud customers. While Copilot for Individuals provides individual users access to Copilot, Copilot for Business makes it easier for companies to purchase and manage licenses for their entire team. If you're considering investing in GitHub Copilot for your team, here's what you need to know.
> 
> ## What is GitHub Copilot? 🤖
> It is a cloud-based artificial intelligence tool developed by GitHub and OpenAI to assist users of Visual Studio, Visual Studio Code, Neovim, and JetBrains by suggesting lines of code and whole functions instantly. You can learn more about GitHub Copilot from the following resources:
> 
> * [An introduction to GitHub Copilot](https://github.com/orgs/community/discussions/39086)
> * [GitHub Copilot Research Recitation](https://github.blog/2021-06-30-github-copilot-research-recitation/)
> * [How GitHub Copilot helps improve developer productivity](https://github.blog/2022-07-14-research-how-github-copilot-helps-improve-developer-productivity/)
> * [8 things you didn’t know you could do with GitHub Copilot](https://github.blog/2022-09-14-8-things-you-didnt-know-you-could-do-with-github-copilot/)
> 
> ## The friction between software development teams and businesses 💼
> Businesses need to make money to survive, pay their employees, and continue paying for the technology that supports their applications. Investors expect a return on investment, and customers expect features that meet their needs. In an ideal world, software developers would have the time and resources to write, test, refactor, and deploy code at their own pace. But the reality is that businesses need features delivered quickly. This can lead to pressure on software developers to move faster, but to the non-developers it can feel like software developers aren’t moving fast enough.
> 
> ### Why are software developers taking so long to finish features and fix bugs? ⏰
> Software development is hard. It’s more than writing code, but many times people, including software engineers, don’t realize that. Sometimes as software developers, we struggle with:
> 
> * Estimating the time it will take to complete a task
> * Envisioning all the problems they may encounter while coding
> * Requesting extra time to understand the complexity of an issue before agreeing to an unreasonable deadline
> 
> Many times in my career, business-centric coworkers like salespeople or product managers have told me, “Here’s the feature we want, but it shouldn’t take you too long.” Spoiler alert: I hate to admit it, but it always takes me longer than expected. Sometimes, it may seem like I just have to update one word or change one picture. However, the request might mean I have to venture into an unfamiliar part of the codebase, learn a new tool, or make a change that could potentially disrupt the entire system. I come from the startup world where we’ve adopted the “Move fast and break things” motto. Realistically, my software development process often looks like this:
> 
> * Researching possibilities
> * Designing the feature
> * Reading documentation
> * Starting to code while referring to documentation
> * Realizing that the initial plan won't work and starting over
> * Adding workarounds to meet the deadline
> * Rushed testing and deployment
> * Fixing bugs after the code breaks in production
> 
> This cycle of coding, testing, and fixing can lead to burnout and decreased motivation to continue building skills and contributing to the company's success.
> 
> ### How to keep software developers happy at your company 😇
> Software developers want to feel valued and have the opportunity to continue building their skills.
> 
> Helping software developers improve their skills is a compelling way for companies to demonstrate their appreciation for developers. To improve their coding skills, software engineers consume educational content, build side projects, attend conferences, and more. However, with a heavy workload, finding time to do anything outside of work-related tasks can be challenging. Many companies provide a percentage of learning time as a benefit for software engineers, but it’s hard to use this benefit when the engineer’s learning time is not accounted for within project deadlines.
> 
> In addition to factoring in a better schedule to balance workload and learning, companies can help reclaim developers upskilling time with Copilot for Business.
> 
> ### How GitHub Copilot can help your team 🤝
> GitHub Copilot helps software developers write code faster by reducing the time spent sifting through documentation, writing boilerplate code, and trying to remember syntax. With GitHub Copilot, software developers can have more time to tap into their passion for coding. Overall, this helps businesses retain investors, customers, and most importantly, software developers. (Yes, you can always replace software developers, but software developers with legacy knowledge of the codebase are invaluable to a company’s success.)
> 
> #### Less time reading documentation 📚
> In the example below, [Chris Sev](https://twitter.com/chris__sev) explains that instead of spending too much time reading documentation and context-switching, GitHub Copilot led him in the right direction.
> 
> > copilot just saved me 547 minutes of reading stripe docs [pic.twitter.com/f1JuRYq32i](https://t.co/f1JuRYq32i)— Chris Sev (@chris__sev) [February 8, 2023](https://twitter.com/chris__sev/status/1623452869516042241?ref_src=twsrc%5Etfw)
> 
> #### Less time writing boilerplate code 🧐
> In the example below, [Kelsey Hightower](https://twitter.com/kelseyhightower) used GitHub Copilot to handle writing boilerplate code.
> 
> > After spending a few minutes with GitHub's Copilot I can see the potential. It's autocomplete with way more context about what you're working on.Just a simple web app but Copilot was able to generate 95% of the code for me. Makes you realize just how much boilerplate we write. [pic.twitter.com/rau501Bba6](https://t.co/rau501Bba6)— Kelsey Hightower (@kelseyhightower) [January 24, 2023](https://twitter.com/kelseyhightower/status/1617684526448402434?ref_src=twsrc%5Etfw)
> 
> #### Less time recalling syntax 🧠
> We all forget syntax sometimes. Memorizing code is not the best use of anyone’s time. In the example below, GitHub Copilot helps [Lisa Reiber](https://twitter.com/lisa_reiber) focus on the logic rather than the syntax.
> 
> > Take on learning with [#githubcopilot](https://twitter.com/hashtag/githubcopilot?src=hash&ref_src=twsrc%5Etfw): for standard syntax it helps to focus on the topic I’m learning (sending data from data lake to data warehouse) instead of searching for the 50th time “if name main syntax” 🎉 it’s even a bit faster than typing[#dezoomcamp](https://twitter.com/hashtag/dezoomcamp?src=hash&ref_src=twsrc%5Etfw) [#vscode](https://twitter.com/hashtag/vscode?src=hash&ref_src=twsrc%5Etfw) [pic.twitter.com/Oz607dVPDr](https://t.co/Oz607dVPDr)— Lisa Reiber (@lisa_reiber) [February 3, 2023](https://twitter.com/lisa_reiber/status/1621529827416375296?ref_src=twsrc%5Etfw)
> 
> ### How GitHub Copilot is useful for me – your underrepresented junior engineer ⭐️
> Beyond productivity, GitHub Copilot helps provide me with psychological safety. Psychological safety is the belief that you can speak up, ask questions, make mistakes, and contribute without fear of punishment or humiliation. When employees feel safe to learn, ask questions, contribute, and lead, it boosts the team’s and individual’s overall confidence, morale, and performance. Unfortunately, psychological safety is not always common at work, especially in the tech industry, especially for minorities.
> 
> I’m a young, black immigrant woman with a non-traditional career path as a developer. I don’t see many people that look like me in my field. I’m passionate about coding, but my first two years working as a Software Engineer were stressful. After I graduated from a coding bootcamp in 2018, I excitedly landed my first software engineering role, but the excitement gradually transformed into anxiety. Asking my teammates for help often led to my teammates questioning my abilities, saying things like:
> 
> * “I thought you would already know this.”
> * “One day, you will learn how to think like a software engineer.”
> * “I know you’re trying to prove yourself as a minority.”
>   I felt so much self-doubt and anxiety from these experiences that it took me hours to open Visual Studio Code. I almost quit coding.
> 
> Discovering GitHub Copilot restored my love for coding. I still ask my teammates questions because I am a collaborative engineer. However, GitHub Copilot acts as my non-judgemental pair programmer, guiding me in the right direction and providing multiple suggestions as I use an API or tool for the first time.
> 
> To me, Copilot for Business can be an investment in empowering your team’s newest developer.
> 
> ### How GitHub Copilot makes coding more accessible 💪
> GitHub Copilot improves the developer experience for engineers with challenges we often overlook such as visual impairment or physical dexterity. The GitHub Next team is a group of researchers and engineers focused on investigating the future of software development, and they developed a hands-free voice-activated AI programmer called “Hey, GitHub.”
> 
> ### Are there any privay concerns I should be aware of? 🕵️‍♂️
> We’ve addressed the privacy and data concerns by implementing the following features:
> 
> * With Copilot for Business, GitHub won’t retain code snippets, store, or share your code regardless if the data is from public repositories, private repositories, non-GitHub repositories, or local files.
> * Additionally, we built a [filter](https://docs.github.com/en/copilot/configuring-github-copilot/configuring-github-copilot-settings-on-githubcom) to help detect and suppress GitHub Copilot suggestions that contain code that matches public code on GitHub. It checks code suggestions with its surrounding code for matches or near matches (ignoring whitespace) against public code on GitHub of about 150 characters. If there is a match, the suggestion will not be shown to you.
> * We are currently building a feature that will provide a reference for suggestions that resemble public code on GitHub so that you can make a more informed decision about whether and how to use that code, as well as explore and learn how that code is used in other projects.
> * We built an AI-based vulnerability prevention system that blocks insecure coding patterns in real-time to make GitHub Copilot suggestions more secure. Our model targets the most common vulnerable coding patterns, including hardcoded credentials, SQL injections, and path injections.
> 
> To learn more about the improvements we’ve made, take a look at the following resources:
> 
> * [GitHub Copilot for Business is now available](https://github.blog/2023-02-14-github-copilot-for-business-is-now-available/)
> * [GitHub Copilot Frequently Asked Questions](https://github.com/features/copilot)
> 
> ### Sign up 🎉
> It’s up to you to decide if Copilot for business is right for your team, but if you’re excited to empower your team, [sign up here](https://github.com/features/copilot)!
> 
> ### Any questions? 💡
> If you have questions, feel free to ask in our [Community discussion board](https://github.com/orgs/community/discussions)

https://github.com/orgs/community/discussions/49422#discussion-4937896
``

**Comment by Billycarter1224** [comment]
If you are planning custom training, a dedicated support team, and more robust security and compliance features,[ yes](https://kekesbreakfastcafes.com/). Your team should use Copilot. Thanks for appreciating, guys.

**Reply by Josephus1123** [reply]
[Uploading bugreport-guamna_trac-RZAS31.Q2-146-14-18-2023-02-23-21-11-41-dumpstate_log-20506.txt…]()

**Reply by Josephus1123** [reply]
Im ready

**Reply by blackgirlbytes** [reply]
I agree @Billycarter1224

**Comment by edugarcam** [comment]
Hi! For our bussiness the most important question is if copilot can learn about our work environment.

How we make a request handler, a function, how our project is organized, what is our database structure, what is the common language for our bussiness,... this is still my most importate doubt, can copilot learn the language we use in our bussiness?

**Reply by itdir** [reply]
It definitely does not. I've seen it make the same mistakes 3 times in a row, even with me providing correct information every time.

**Comment by punkproger** [comment]
Hello
I want to push my company to use GitHub Coplitot, but we use Bitbucket for git repositories system
Can we somehow use GitHub Coplitot for Business, but without GitHub?

**Comment by michelvd** [comment]
I see in the "business" plan: " Industry-leading privacy".
Should I conclude that I don't get " Industry-leading privacy" for my individual plan?

**Reply by raflyano-nur** [reply]
_Oijnk**jjuikkkk**_

**Reply by raflyano-nur** [reply]
kkkk

**Comment by psychowood** [comment]
`With Copilot for Business, GitHub won’t retain code snippets, store, or share your code regardless if the data is from public repositories, private repositories, non-GitHub repositories, or local files.`

@blackgirlbytes is this applicable to Copilot Chat, too, especially since it is in Pre-Release? I'm particularly interested in the [4. Data Collection and Usage](https://docs.github.com/en/site-policy/github-terms/github-copilot-pre-release-terms#4-data-collection-and-usage) point.

I'm talking about a Copilot for Business subscription.

Thanks

**Comment by clareqinty89** [comment]
It sounds like you're weighing your options for improving your team's development process. While GitHub Copilot offers some exciting features, integrating it with Azure DevOps might require some extra consideration. Have you explored how [Monday project management](https://www.enable.services/) could fit into the mix? It might offer some solutions to streamline your workflow effectively.
